packages/core/src/routes/phrase.ts

```diff
@@ -0,0 +1,130 @@
+import { LanguageKey, languageKeyGuard } from '@logto/core-kit';
+import resource from '@logto/phrases-ui';
+import en from '@logto/phrases-ui/lib/locales/en';
+import { CustomPhrases, Translation } from '@logto/schemas';
+import { adminConsoleApplicationId, adminConsoleSignInExperience } from '@logto/schemas/lib/seeds';
+import { notFalsy } from '@silverhand/essentials';
+import deepmerge from 'deepmerge';
+import { ResourceLanguage } from 'i18next';
+import { errors, Provider } from 'oidc-provider';
+
+import detectLanguage from '@/i18n/detect-language';
+import koaGuard from '@/middleware/koa-guard';
+import { findAllCustomLanguageKeys, findCustomPhraseByLanguageKey } from '@/queries/custom-phrase';
+import { findDefaultSignInExperience } from '@/queries/sign-in-experience';
+import assertThat from '@/utils/assert-that';
+
+import { AnonymousRouter } from './types';
+
+const getLanguageInfo = async (applicationId: unknown) => {
+  if (applicationId === adminConsoleApplicationId) {
+    return adminConsoleSignInExperience.languageInfo;
+  }
+
+  const { languageInfo } = await findDefaultSignInExperience();
+
+  return languageInfo;
+};
+
+const isBuiltInLanguageKey = (key: string): key is LanguageKey =>
+  Object.keys(resource).includes(key) && languageKeyGuard.safeParse(key).success;
+
+const isFullyTranslated = (fullTranslation: Translation, targetTranslation: Translation) => {
```

Comment:
deep merge will do all the works, deep merge(custom, build in target, en custom, build in en);